core/common/scale.cpp

```diff
@@ -0,0 +1,493 @@
+        }
+
+        const uint8_t flag = first_byte.value() & 0b11;
+
+        size_t number = 0;
+
+        switch (flag) {
+          case 0b00: {
+            number = static_cast<size_t>(first_byte.value() >> 2);
+            break;
+          }
+
+          case 0b01: {
+            auto second_byte = nextByte();
+            if (!second_byte.has_value()) {
+              return {};
+            }
+
+            number = (static_cast<size_t>(first_byte.value() & 0b11111100)
+                      + static_cast<size_t>(second_byte.value()) * 256)
+                >> 2;
+            break;
+          }
+
+          case 0b10: {
+            number = first_byte.value();
+            size_t multiplier = 256;
+            if (!hasMore(3)) {
+                // not enough data to decode integer
+                return {};
+              }
+
+            for (int i = 0; i < 3; ++i) {
+                // we assured that there are 3 more bytes,
+                // no need to make checks in a loop
+              number += nextByte().value() * multiplier;
+              multiplier = multiplier << 8;
+            }
+            number = number >> 2;
+            break;
+          }
+
+          case 0b11: {
+            uint8_t m = static_cast<uint8_t>((first_byte.value() >> 2) + 4);
+            if (!hasMore(m)) {
+              // not enough data to decode integer
+              return {};
+            }
+
+            BigInteger multiplier{1};
+            BigInteger length = 0;
+            for (uint8_t i = 1; i < m; ++i) {
+                // we assured that there are m more bytes,
+                // no need to make checks in a loop
+              length += nextByte().value() * multiplier;
+              multiplier << 8;
+            }
+
+            return {{flag, length}};  // special case
+          }
+
+          default:
+            break;  // unreachable code
+        }
+
+        return {{flag, BigInteger{number}}};
+      }
+
+      expected::Result<ByteArray, DecoderError> Decoder::decodeBytes() {
+        if (!hasMore())
+          return {};
+
+        auto result = decodeInteger();
+        if (!result.has_value())
+          return expected::Error{DecoderError::kNotEnoughData};
+
+        BigInteger number = std::move(result.value().second);
+
+        if (number
+            >= std::numeric_limits<std::vector<uint8_t>::size_type>::max() / 2)
+          return expected::Error{DecoderError::kValueIsTooBig};
+
+        int64_t amount = static_cast<int64_t>(number);
+        if (bytesLeft() < amount)
+          return expected::Error{DecoderError::kNotEnoughData};
+
+        ByteArray bytes;
+        bytes.reserve(amount);
+
+        auto begin = current_iterator_;
+        auto end = begin + amount;
+
+        current_iterator_ = end;
+        bytes_left_ -= amount;
+
+        std::copy(begin, end, bytes.end());
+
+        return expected::Result<ByteArray, DecoderError>::ValueType{
+            std::move(bytes)};
```

Comment:
I think `bytes` variable is located on the stack and moving it will not make code more efficient

Reply:
Content of a vector is on the heap.